explorer/interpreter/type_checker.cpp

```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+  void Add(std::vector<AbstractPattern> pattern_vector) {
+    assert(matrix_.empty() || matrix_[0].size() == pattern_vector.size());
+    matrix_.push_back(std::move(pattern_vector));
+  }
+
+  // Determine whether the given pattern vector is useful: that is, whether
+  // adding it to the matrix would allow any more values to be matched.
+  auto IsUseful(llvm::ArrayRef<AbstractPattern> pattern) const -> bool {
+    if (matrix_.empty()) {
+      return true;
+    }
+
+    assert(pattern.size() == matrix_[0].size());
+    if (matrix_[0].empty()) {
+      return false;
+    }
+
+    switch (pattern[0].kind()) {
+      case AbstractPattern::Wildcard: {
+        auto discrim = FirstColumnDiscriminators();
+        if (!discrim.any_missing) {
+          for (auto found : discrim.found) {
+            if (Specialize(found).IsUseful(*SpecializeRow(pattern, found))) {
+              return true;
+            }
+          }
+          return false;
+        }
+        return Default().IsUseful(pattern.slice(1));
+      }
+
+      case AbstractPattern::Compound: {
+        DiscriminatorInfo discrim = {
+            .discriminator = pattern[0].discriminator(),
+            .size = pattern[0].NumElements()};
+        return Specialize(discrim).IsUseful(*SpecializeRow(pattern, discrim));
+      }
+
+      case AbstractPattern::Primitive: {
+        return Specialize(pattern[0].value()).IsUseful(pattern.slice(1));
+      }
+    }
+  }
+
+ private:
+  // Information about a constructor for a compound type.
+  struct DiscriminatorInfo {
+    // For an alternative, the name. Otherwise, empty.
+    std::string_view discriminator;
+    // The number of elements. For a tuple, the size. Always 1 for an
+    // alternative.
+    int size;
+  };
+
+  struct DiscriminatorSet {
+    std::vector<DiscriminatorInfo> found;
+    bool any_missing;
+  };
+
+  // Find the discriminators used by the first column and check whether we
+  // found all of them.
+  auto FirstColumnDiscriminators() const -> DiscriminatorSet {
+    std::set<std::string_view> discrims;
+    std::optional<const ChoiceType*> choice;
+
+    for (auto& row : matrix_) {
+      assert(!row.empty());
+      switch (row[0].kind()) {
+        case AbstractPattern::Wildcard:
+          continue;
+        case AbstractPattern::Compound: {
+          if (auto* tuple = dyn_cast<TupleValue>(&row[0].type())) {
+            // If we find a tuple match, we've found all constructors (there's
+            // only one!) and none were missing.
+            return {
+                .found = {{.discriminator = {},
+                           .size = static_cast<int>(tuple->elements().size())}},
+                .any_missing = false};
+          }
+          discrims.insert(row[0].discriminator());
+          choice = &cast<ChoiceType>(row[0].type());
+          break;
+        }
+        case AbstractPattern::Primitive: {
+          // TODO: We assume that primitive value matches are always incomplete
```

Comment:
Can you add a simple test with `case 1`, `case 2`, etc just to make sure it's working as intended?

Maybe do something with `case 1` repeated?

Reply:
Done.